### ISSUE TITLE
more Option functions to consider

### ISSUE THREAD

**Issue by ryanguill** [iwburns/nullshield#19]
There are lots of different implementations of an Option/Maybe monad out there and of course there are tons of different features.  Just wanted to collect these for consideration.

- [x] `forEach(func: () => void) : void` for side effects.  Called for a `some`, basically a no-op for a `none`.
- [x] `filter<U>(condition: (val: T) => boolean) : OptT<U>` for `none` returns a `none`, for a `some`, val is provided to the condition and if it returns true then the same `some` with the same value is returned, otherwise a `none`. 
- [x] `contains(condition: (val: T) => boolean) : boolean` basically like `filter`, but will return true for a `some` that passes the condition and `none` in all other cases.

**Comment by iwburns** [comment]
Replying on mobile, so this will probably be short for now.

I'm definitely down with these.  I thought I had filter in there already, but apparently not.

I'm not sold on the names `forEach` and `contains`, but I might come around to them.

My thought for the former is we could alternatively call it  `ifSome` so that we can also have `ifNone` (but maybe that's not necessary?).

For `contains` it almost seems like we'd want to it set up like this `contains(val: T): boolean` and then have `containsBy` which is what you have above (again, maybe not necessary?).

**Comment by iwburns** [comment]
So @ryanguill (and @brycehipp if you're interested in discussing), I've finally had some more time to do a bit more detailed write up on the above and I'd like to get your opinion(s) on it.

#### First ####
A small question about the signature of `filter`: does it really need to be generic over `U`?  It seems like it could just return `OptT<T>`, but maybe I'm missing something.

#### Second ####
I'm curious if `forEach(func: () => void): void` should be split into the following two functions or not:
* `ifSome<T>(func: (val: T) => void): void` (tweaked a bit by passing in `val`).
* `ifNone(func: () => void): void`

My first thought here is it would be nice to have an "inverse" of `forEach` in the case where you only want to execute a function if something is a `None`.  The reason for the name change is purely for symmetry and because I couldn't think of anything that sounded better.

However, I'm not sure if `ifNone` is actually helpful or not because the following options don't really seem different enough to matter (although that is certainly just my opinion):
```
const none = OptionT.none();

none.ifNone(someOtherFunc);
```
```
const none = OptionT.none();

if (none.isNone()) {
  someOtherFunc();
}
```

#### Third ####
I'm not familiar enough with existing implementations of `contains`, but to me it would make more sense for `contains` to just be comparing the values directly. In that case you don't need a `condition` function, you just need a `val` to compare with.  For more complicated comparisons though (maybe Objects with keys that you're comparing), I could see the need for a custom comparison function.

I would think splitting up `contains` into the following would make the most sense:
* `contains(val: T): boolean`
* `containsBy(condition: (val: T) => boolean): boolean` (not 100% sold on the name).

This way, you could write both of the following and I feel like they mostly make sense:
```
const maybeOne = OptionT.some(1);
if (maybeOne.contains(1)) {
  // do stuff
}
```
```
const maybeObj = OptionT.some({ foo: 'bar' });

const fooIsBar = (val) => val.foo === 'bar';

if (maybeObj.containsBy(fooIsBar)) {
  // do stuff
}
```
I don't really _love_ the way that last example reads, but it might just be the case that this is a poor example and "real-world" code would read fine here.

I'd love to get your thoughts / ideas on this and discuss the best way forward if / when you have the time.

**Comment by ryanguill** [comment]
> A small question about the signature of filter: does it really need to be generic over U? It seems like it could just return OptT<T>, but maybe I'm missing something.

No, I think youre right.  

> I'm curious if forEach(func: () => void): void should be split into the following two functions or not...

The point of `forEach` (and these other functions in this request) is more about being able to treat the option as a collection - map / filter / forEach are the staples there for transform / exclude / do some side-effect-y thing.  I'm also trying to match api's with what im mostly familiar with.

> I'm not familiar enough with existing implementations of contains, but to me it would make more sense for contains to just be comparing the values directly.

I think you're thinking of `contains` like its an `equals` - im really trying to use it more like a companion to `filter` - but instead of getting the value if its passes I just get a boolean if it will pass or not.  I think there might be a use for an `equals` - and I admit that this function could be used that way, but I don't think it was my intended use case. 

Consider if you want to check if a value is within a range, or within a set of accepted values.

` maybe.contains(x => x != 0);`
`maybe.contains(x => x 0 <= x <= 100);`
```
const s = new Set(['A','B','C']);
maybe.contains(x => s.has(x))
```

**Comment by iwburns** [comment]
> The point of `forEach` ...

I'm with you on that.  I'd like to follow that API as much as possible as well.  Further, I'm not really sure `isNone` is _that_ useful.  If we (or others) find a need for it later though, it'll be easy enough to add something like that.

> I think you're thinking of `contains` like its an `equals` ...

My only argument here is that `equals` would make the most sense looking like this:

`equals(other: OptT<T>): boolean`

instead of:

`equals(val: T): boolean`.

I actually think the former could be useful, so I'd rather not use the latter for `equals`.

At this point I think we're just running into the "naming things is hard" problem. I think we both agree that both versions of this function are useful (correct me if I'm wrong of course), but coming up with a name for both versions that makes sense isn't easy.

---

With all of that in mind, here's what I'm thinking so far:

I'm good with these:
* `filter(condition: (val: T) => boolean): OptT<T>`
* `forEach(func: () => void): void`
* `equals(other: OptT<T>): boolean`

Not sure on these:
* `contains(condition: (val: T) => boolean): boolean`  (also could be called `containsBy`)
* `hasValue(val: T): boolean` (also could be called `contains`)

How do those options sound to you?

**Comment by iwburns** [comment]
ping @ryanguill I know you've been busy, just want your opinion on the above if/when you have time.

**Comment by iwburns** [comment]
@ryanguill I just merged #25 which takes care of this in the way that I described above.  Let me know what you think if you have time.

I went with the best names I could think of at the time, but I'm definitely willing to talk about better names/other behaviors if you have any thoughts along those lines.

**Comment by iwburns** [comment]
closing this as I believe it has been resolved.
